File: case-studies/02-payment-system/java/PaymentCoordinator.java
// Scenario: idempotent payment creation in a payment system
// Demonstrates: idempotency lookup, provider interaction, and response caching
// Trade-off: simplicity versus storage overhead for idempotency records

package case_studies.payment;

public class PaymentCoordinator {
    private final IdempotencyStore idempotencyStore;
    private final PaymentStore paymentStore;
    private final PaymentProvider provider;

    public PaymentCoordinator(IdempotencyStore idempotencyStore, PaymentStore paymentStore, PaymentProvider provider) {
        this.idempotencyStore = idempotencyStore;
        this.paymentStore = paymentStore;
        this.provider = provider;
    }

    public PaymentResponse handleCreate(PaymentRequest request) {
        IdempotencyRecord existing = idempotencyStore.lookup(request.getIdempotencyKey());
        if (existing != null) {
            return existing.getResponse();
        }

        PaymentRecord record = paymentStore.create(request);
        PaymentResponse response = provider.authorize(record);
        idempotencyStore.save(request.getIdempotencyKey(), response);
        return response;
    }
}

File: case-studies/02-payment-system/go/payment_coordinator.go
// Scenario: idempotent payment creation in a payment system
// Demonstrates: idempotency lookup, provider authorization, and response caching
// Trade-off: storage overhead versus duplicate-charge safety

package payment

type PaymentCoordinator struct {
	IdempotencyStore IdempotencyStore
	PaymentStore     PaymentStore
	Provider         PaymentProvider
}

func (c *PaymentCoordinator) HandleCreate(req PaymentRequest) (PaymentResponse, error) {
	if rec, err := c.IdempotencyStore.Lookup(req.IdempotencyKey); err == nil && rec != nil {
		return rec.Response, nil
	}

	payment, err := c.PaymentStore.Create(req)
	if err != nil {
		return PaymentResponse{}, err
	}

	response, err := c.Provider.Authorize(payment)
	if err != nil {
		return PaymentResponse{}, err
	}

	if err := c.IdempotencyStore.Save(req.IdempotencyKey, response); err != nil {
		return PaymentResponse{}, err
	}

	return response, nil
}
